### ISSUE TITLE
How do I pass a pointer to the _mm_cmpestrm instrinsic?

### ISSUE THREAD

**Issue by shepmaster** [rust-lang-nursery/stdsimd#360]
I'm attempting to port [Jetscii](https://github.com/shepmaster/jetscii) to stdsimd, but I'm unclear on how the `_mm_cmpestrm` intrinsic deals with pointers.

This is the current inline assembly:

```rust
        let matching_bytes;

        asm!("movlhps $2, $1
              pcmpestrm $$0, ($3), $1"
             : // output operands
             "={xmm0}"(matching_bytes)
             : // input operands
             "x"(self.needle_lo),
             "x"(self.needle_hi),
             "r"(ptr),
             "{rdx}"(offset + len), // saturates at 16
             "{rax}"(self.count as u64)
             : // clobbers
             "cc"
             : // options
        );

        matching_bytes
```

From the Intel x86_64 Reference Manual (emphasis mine):

> Each string fragment is represented by two values. The first value is an xmm (**or possibly m128 for the second operand**) which contains the data elements of the string (byte or word data). The second value is stored in an input length register. The input length register is EAX/RAX (for xmm1) or **EDX/RDX (for xmm2/m128)**. The length represents the number of bytes/words which are valid for the respective xmm/m128 data.

My current (non-working) implementation attempts to just cram the pointer into an `__m128i`, but that's highly unlikely to be correct 😉 :

```rust
        use std::simd::u64x2;
        use std::arch::x86_64::{_mm_cmpestrm, _mm_set_epi64x};

        let needle = _mm_set_epi64x(self.needle_hi as _, self.needle_lo as _);

        let haystack = _mm_set_epi64x(0, ptr as _);

        let r = _mm_cmpestrm(needle, self.count as _, haystack, (offset + len) as _, 0);
        let r: u64x2 = ::std::mem::transmute(r);
        r.extract(0)
```

**Comment by alexcrichton** [comment]
Intel's [documentation](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_cmpestrm&expand=811,811) seems to indicate that the two arguments are "packed strings" which sort of implies to me that they're not pointers but rather 128bits of contiguous string loaded into registers. Haven't not used these intrinsics before I'm not entirely sure, but maybe that'd help?

**Comment by alexcrichton** [comment]
I also came across https://github.com/WojciechMula/simd-string which may help

**Comment by gnzlbg** [comment]
I think you need to use ‘some128vec.load_aligned/unaligned’ to load them
into a std::simd vector and then into_bits __m128.

**Comment by Cocalus** [comment]
```rust
let haystack = _mm_loadu_m128i(ptr as *const __m128i);
```

Note that if you don't have 16 bytes of memory at the ptr, then there's a rare segfault hiding in there. This true of the asm variant as well. For example if ptr only points to 15 bytes of valid memory and the final byte is on the next page, and that page isn't mapped into your process or isn't doesn't have readable protection then you'll have a segfault. The simplest solution is to just pad 15 bytes onto what ever ptr points to. To check for this kind of thing memmap and libc::mprotect with libc::PROT_NONE can be used to put an inaccessible page a known position then you can put the end of your string just before the edge of it and test it.

You can also try flipping the inputs to _mm_cmpestrm, different assembly formats flip them, which makes some docs or examples confusing.

Many of the x86 assembly instructions can take a memory address instead of a register. I forget whether they need to be aligned in general. But they do not need be aligned for cmpestrm and friends. The intrinsics don't expose that, the hope is that either the compiler can optimize out the load and use the memory form, or it doesn't affect performance (other than wasting some instruction cache). 

Note the cmpistrm variation (3 cycle latency) is faster then cmpestrm (5 cycle latency), but that requires zero terminated strings.

**Comment by shepmaster** [comment]
> Note that if you don't have 16 bytes of memory at the ptr, then there's a rare segfault hiding in there.

Indeed, which is why [I have tests for that case](https://github.com/shepmaster/jetscii/blob/7b3e62ce3e1ec85c16c14a45ad74cd91d4aa1bb6/src/lib.rs#L808-L849) :-)

**Comment by shepmaster** [comment]
> I think you need to use ‘some128vec.load_aligned/unaligned’ to load them
> into a std::simd vector and then into_bits __m128.

Yeah, that's what I'm currently doing, but it *feels* inefficient. I know that the instruction has to move data from memory either way, but it seems as if letting the instruction know that it's reading from memory would allow it to schedule the instruction better. When I get my tests passing, I'll attempt benchmarking.

**Comment by shepmaster** [comment]
Rough benchmarking seems to indicate that explicit loading from memory doesn't have a negative effect. It might even have a benefit. I'm on a train where power is going in and out, so take it with a gain of salt...

```
 name                                      control ns/iter        variable ns/iter       diff ns/iter   diff %  speedup
 bench::space_asciichars                   892,740 (5872 MB/s)    945,273 (5546 MB/s)          52,533    5.88%   x 0.94
 bench::space_asciichars_as_pattern        878,239 (5969 MB/s)    902,850 (5807 MB/s)          24,611    2.80%   x 0.97
 bench::space_asciichars_macro             934,715 (5609 MB/s)    913,505 (5739 MB/s)         -21,210   -2.27%   x 1.02
 bench::space_find_byte                    3,475,138 (1508 MB/s)  3,811,689 (1375 MB/s)       336,551    9.68%   x 0.91
 bench::space_find_char                    662,240 (7916 MB/s)    670,797 (7815 MB/s)           8,557    1.29%   x 0.99
 bench::space_find_char_set                10,634,048 (493 MB/s)  10,358,239 (506 MB/s)      -275,809   -2.59%   x 1.03
 bench::space_find_closure                 10,278,144 (510 MB/s)  10,096,480 (519 MB/s)      -181,664   -1.77%   x 1.02
 bench::space_find_string                  7,865,016 (666 MB/s)   7,162,629 (731 MB/s)       -702,387   -8.93%   x 1.10
 bench::substring_as_pattern               1,005,900 (5212 MB/s)  1,003,825 (5222 MB/s)        -2,075   -0.21%   x 1.00
 bench::substring_find                     1,603,561 (3269 MB/s)  1,655,691 (3166 MB/s)        52,130    3.25%   x 0.97
 bench::substring_with_cached_searcher     980,867 (5345 MB/s)    996,624 (5260 MB/s)          15,757    1.61%   x 0.98
 bench::substring_with_created_searcher    1,021,119 (5134 MB/s)  989,778 (5297 MB/s)         -31,341   -3.07%   x 1.03
 bench::xml_delim_3_asciichars             918,994 (5705 MB/s)    879,762 (5959 MB/s)         -39,232   -4.27%   x 1.04
 bench::xml_delim_3_asciichars_as_pattern  905,883 (5787 MB/s)    882,446 (5941 MB/s)         -23,437   -2.59%   x 1.03
 bench::xml_delim_3_asciichars_macro       884,363 (5928 MB/s)    882,594 (5940 MB/s)          -1,769   -0.20%   x 1.00
 bench::xml_delim_3_find_byte_closure      2,233,100 (2347 MB/s)  2,164,688 (2422 MB/s)       -68,412   -3.06%   x 1.03
 bench::xml_delim_3_find_char_closure      14,290,244 (366 MB/s)  10,605,283 (494 MB/s)    -3,684,961  -25.79%   x 1.35
 bench::xml_delim_3_find_char_set          17,576,367 (298 MB/s)  18,576,023 (282 MB/s)       999,656    5.69%   x 0.95
 bench::xml_delim_5_asciichars             889,703 (5892 MB/s)    892,973 (5871 MB/s)           3,270    0.37%   x 1.00
 bench::xml_delim_5_asciichars_as_pattern  894,272 (5862 MB/s)    887,115 (5910 MB/s)          -7,157   -0.80%   x 1.01
 bench::xml_delim_5_asciichars_macro       903,936 (5800 MB/s)    885,327 (5921 MB/s)         -18,609   -2.06%   x 1.02
 bench::xml_delim_5_find_byte_closure      2,259,781 (2320 MB/s)  2,191,320 (2392 MB/s)       -68,461   -3.03%   x 1.03
 bench::xml_delim_5_find_char_closure      7,731,400 (678 MB/s)   6,980,080 (751 MB/s)       -751,320   -9.72%   x 1.11
 bench::xml_delim_5_find_char_set          21,427,295 (244 MB/s)  20,932,433 (250 MB/s)      -494,862   -2.31%   x 1.02
```

**Comment by shepmaster** [comment]
Now to see if there's some way of implementing this:

```rust
        asm!("# Rebuild the needle from parts
              # - move low word of $2 to high word of $1
              movlhps $3, $2
1:
              # pcmpestri control byte, (pointer, offset), needle
              pcmpestri $$0, ($4, $1), $2
              # Carry flag is set when an index is found
              jc 2f
              addq $$16, $1
              subq $$16, $5
              jnz 1b
2:
"
             : // output operands
             "={ecx}"(res),
             "+r"(offset)
             : // input operands
             "x"(self.needle_lo),
             "x"(self.needle_hi),
             "r"(ptr),
             "{rdx}"(len),              // haystack length
             "{rax}"(self.count as u64) // needle length
             : // clobbers
             "rcx",
             "memory",
             "cc"
             : // options
         );
```

**Comment by Cocalus** [comment]
The compiler should be able to mix a _mm_cmpestrm followed by a _mm_cmpestrz with the same inputs into a single pcmpestri instruction. 

See this [stack overflow question](https://stackoverflow.com/questions/40563615/sse-intrinsics-check-zero-flag)

**Comment by gnzlbg** [comment]
@Cocalus 

> Note that if you don't have 16 bytes of memory at the ptr, then there's a rare segfault hiding in there. 

The functions `load_{aligned, unaligned}{_unchecked}` should be helpful here. If you do `load_aligned` and the pointer is not aligned to a 16 byte boundary, then you will get a `panic!` instead of a segfault that you can trace with `RUST_BACKTRACE=1`. If you want to omit the checks, just use the `_unchecked` variants.

**Comment by gnzlbg** [comment]
@shepmaster 

> Yeah, that's what I'm currently doing, but it feels inefficient.

I think that is just a feeling. SIMD vectors are registers. If you want to initialize them to some values, you need to do a load (either from the stack, or main memory). The only way in which this would be inefficient is if you didn't had to do a load at all, for example, because you already have a vector with the values in another register, or because you want to initialize it to a constant value.

**Comment by shepmaster** [comment]
> The functions load_{aligned, unaligned}{_unchecked} should be helpful here.

I'm not sure @Cocalus' point is fully addressed by these. The `PCMPxSTRx` instructions *themselves* handle reading from unaligned memory just fine.

The problem arises when you read from a pointer that is (say) 1 byte before a page boundary. The instruction will then attempt to read 16 bytes, crossing between pages. If the second page is disallowed for a number of reasons, you will get a page fault. 

The `load_unaligned_unchecked` function will have the same problem; the caller has to ensure that the length doesn't cross into the next page. The `_checked` function would have the overhead of checking length on every memory access and will fail if the length is less than 16 bytes, which it very likely will be for a string.

**Comment by gnzlbg** [comment]
@shepmaster there was a discussion here: https://github.com/rust-lang-nursery/stdsimd/issues/199 about using pointer types that required a specific alignment back then when we had a strongly typed API. 

The idea was, in a nutshell, that a strongly-typed wrapper around these wouldn't take a raw pointer but a pointer to a type that must be 16byte aligned. However, if you wanted to create this pointer type from a pointer to a string, you'd either need to call a safe constructor that would do the alignment check, or use an `unsafe` function that just casts / transmutes the pointer. 

The way to safely avoid the check would then be to store your strings in some type that ensures that they are 16byte aligned.

**Comment by alexcrichton** [comment]
@shepmaster it looks like https://github.com/WojciechMula/simd-string/blob/master/strchr.cpp may be adaptable to finding indexes?

**Comment by shepmaster** [comment]
> However, if you wanted to create this pointer type from a pointer to a string, you'd either need to call a safe constructor that would do the alignment check, or use an unsafe function that just casts / transmutes the pointer.
> 
> The way to safely avoid the check would then be to store your strings in some type that ensures that they are 16byte aligned, and that is able to return you a U8_ptr for it internally using the transmute.

This is basically my plan. The working Rust code (from years before stdsimd existed) essentially 16-byte aligns the input string's pointer and then hands it to `PCMPESTRx`, albeit without the benefit of a type. The stdsimd-port is also using that right now, but it's hurting my brain too much to maintain, so I'm going to be moving to a type that can help uphold the invariants.

**Comment by alexcrichton** [comment]
Ok it sounds like enough progress has been made that I think this question is semi-resolved, but if not I can certainly reopen!

**Comment by shepmaster** [comment]
> How do I pass a pointer to the _mm_cmpestrm instrinsic?

TL;DR: you don't. Instead, you need to load from the pointer into a `__m128i` and then pass that to the intrinsic.

You can either load from a cast pointer directly:

```rust
let haystack = _mm_loadu_m128i(ptr as *const __m128i);
```

or via a slice:

```rust
let haystack: &[u8] = /* ... */;
let haystack = u8x16::load_unaligned_unchecked(haystack); // or other load_* functions
let haystack: __m128i = haystack.into_bits();
```
